components/style/values/computed/counters.rs

```diff
@@ -4,11 +4,202 @@
 
 //! Computed values for counter properties
 
+#[cfg(feature = "servo")]
+use computed_values::list_style_type::T as ListStyleType;
+use cssparser::{self, Parser, Token};
+use parser::{Parse, ParserContext};
+use selectors::parser::SelectorParseErrorKind;
+use std::fmt::{self, Write};
+use style_traits::{CssWriter, ParseError, StyleParseErrorKind, ToCss};
+#[cfg(feature = "gecko")]
+use values::generics::CounterStyleOrNone;
 use values::generics::counters::CounterIncrement as GenericCounterIncrement;
 use values::generics::counters::CounterReset as GenericCounterReset;
+#[cfg(feature = "gecko")]
+use values::specified::Attr;
+use values::specified::url::SpecifiedUrl;
+pub use values::specified::{Content, ContentItem};
 
 /// A computed value for the `counter-increment` property.
 pub type CounterIncrement = GenericCounterIncrement<i32>;
 
 /// A computed value for the `counter-increment` property.
 pub type CounterReset = GenericCounterReset<i32>;
+
+impl Content {
+    /// Set `content` property to `normal`.
+    #[inline]
+    pub fn normal() -> Self {
+        Content::Normal
+    }
+
+    #[cfg(feature = "servo")]
+    fn parse_counter_style(
+        input: &mut Parser
+    ) -> ListStyleType {
+        input.try(|input| {
+            input.expect_comma()?;
+            ListStyleType::parse(input)
+        }).unwrap_or(ListStyleType::Decimal)
+    }
+
+    #[cfg(feature = "gecko")]
+    fn parse_counter_style(
+        context: &ParserContext,
+        input: &mut Parser
+    ) -> CounterStyleOrNone {
+        input.try(|input| {
+            input.expect_comma()?;
+            CounterStyleOrNone::parse(context, input)
+        }).unwrap_or(CounterStyleOrNone::decimal())
+    }
+}
+
+impl Parse for Content {
+    // normal | none | [ <string> | <counter> | open-quote | close-quote | no-open-quote |
+    // no-close-quote ]+
+    // TODO: <uri>, attr(<identifier>)
+    fn parse<'i, 't>(
+        context: &ParserContext,
+        input: &mut Parser<'i, 't>
+    ) -> Result<Self, ParseError<'i>> {
+        if input.try(|input| input.expect_ident_matching("normal")).is_ok() {
+            return Ok(Content::Normal)
+        }
+        if input.try(|input| input.expect_ident_matching("none")).is_ok() {
+            return Ok(Content::None)
+        }
+        if input.try(|input| input.expect_ident_matching("-moz-alt-content")).is_ok() {
+            #[cfg(feature = "gecko")]
```

Comment:
I think this checking is only used for `gecko`.
Will it be better to be written like following?

```rust
#[cfg(feature = "gecko")] {
  if input.try(|input| input.expect_ident_matching("-moz-alt-content")).is_ok() {
    return Ok(Content::MozAltContent)
  }
}
```

Reply:
Unfortunately, no. We can't use `#[cfg]` attributes for `if` blocks:
```Rust
error: attributes are not yet allowed on `if` expressions                                                                                                                                                                                                                                                    
  --> components/style/values/computed/counters.rs:72:9                                                                                                                                                                                                                                                      
   |                                                                                                                                                                                                                                                                                                         
72 |         #[cfg(feature = "gecko")]                                                                                                                                                                                                                                                                       
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                                             
error: aborting due to previous error                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                             
error: Could not compile `style`.
```

Reply:
Oh, I think we could use the `cfg!` macro like so:
```Rust
if cfg!(feature = "gecko") {
  if input.try(|input| input.expect_ident_matching("-moz-alt-content")).is_ok() {
    return Ok(Content::MozAltContent)
  }
}
```
This actually compiles.

Reply:
It won't compile on servo I guess, because `MozAltContent` is not defined here.

I haven't taken an in-depth look at this, but generally you can wrap it in a block like:

```rust
#[cfg(feature = "gecko")]
{
    if ... {
        return Ok(Content::MozAltContent);
    }
}
```

We use that pattern a lot in a bunch of places.

Reply:
@emilio Thanks for the tip! It should be wrap in a block!

Reply:
Oh, I see. I didn't wrap it in a block indeed, so it didn't compile.

@CYBAI You were right all along, didn't see the curly braces :).